feat(parsers): Add kitty

OXY2DEV opened the pull request:
<!-- 
  Before proceeding, make sure you have read https://github.com/nvim-treesitter/nvim-treesitter/blob/main/CONTRIBUTING.md!
  Make sure to fill out all fields and read the checklist at the end.
-->

# Name of language

[kitty](https://sw.kovidgoyal.net/kitty/conf/)

<details>
<summary>Representative code sample</summary>

```kitty
font_family      monospace
bold_font        auto

map alt+1 disable_ligatures_in active always
map kitty_mod+s  paste_from_selection
map shift+insert paste_from_selection

modify_font strikethrough_position 2px
box_drawing_scale 0.001, 1, 1.5, 2
undercurl_style thin-sparse

cursor #cccccc
cursor_blink_interval 10 ease-in
cursor_blink_interval 10 steps(10, end)

mouse_map left click ungrabbed mouse_handle_click selection link prompt
mouse_map shift+left click grabbed,ungrabbed mouse_handle_click selection link prompt
mouse_map ctrl+shift+left release grabbed,ungrabbed mouse_handle_click link
mouse_map ctrl+shift+alt+left triplepress ungrabbed,grabbed mouse_selection line_from_point

env VAR1=a
env VAR2=${HOME}/${VAR1}/b
```

</details>

## Parser repo

https://github.com/OXY2DEV/tree-sitter-kitty

<details>
<summary>Parsed tree for code sample</summary>

```query
(configuration_file [0, 0] - [22, 0]
  (font_option [0, 0] - [0, 26]
    value: (font_value [0, 17] - [0, 26]
      (string [0, 17] - [0, 26])))
  (font_option [1, 0] - [1, 21]
    value: (font_value [1, 17] - [1, 21]
      (string [1, 17] - [1, 21])))
  (keyboard_shortcut [3, 0] - [3, 44]
    sequence: (key_sequence [3, 4] - [3, 9]
      (modified_key [3, 4] - [3, 9]
        (alt [3, 4] - [3, 7])
        (key [3, 8] - [3, 9])))
    action: (key_action [3, 10] - [3, 44]
      (disable_ligatures_in [3, 10] - [3, 44]
        target: (ligature_target [3, 31] - [3, 37])
        type: (ligature_disabled [3, 38] - [3, 44]))))
  (keyboard_shortcut [4, 0] - [4, 37]
    sequence: (key_sequence [4, 4] - [4, 15]
      (special [4, 4] - [4, 13])
      (key [4, 14] - [4, 15]))
    action: (key_action [4, 17] - [4, 37]
      (generic_action [4, 17] - [4, 37])))
  (keyboard_shortcut [5, 0] - [5, 37]
    sequence: (key_sequence [5, 4] - [5, 16]
      (modified_key [5, 4] - [5, 16]
        (shift [5, 4] - [5, 9])
        (special [5, 10] - [5, 16])))
    action: (key_action [5, 17] - [5, 37]
      (generic_action [5, 17] - [5, 37])))
  (modify_font [7, 0] - [7, 38]
    type: (font_modification_type [7, 12] - [7, 34])
    value: (pixel [7, 35] - [7, 38]
      (number [7, 35] - [7, 36])))
  (box_drawing_scale [8, 0] - [8, 34]
    value: (scale_value [8, 18] - [8, 34]
      (number [8, 18] - [8, 23])
      (number [8, 25] - [8, 26])
      (number [8, 28] - [8, 31])
      (number [8, 33] - [8, 34])))
  (undercurl_style [9, 0] - [9, 27])
  (color_option [11, 0] - [11, 14]
    name: (color_option_name [11, 0] - [11, 6])
    value: (color [11, 7] - [11, 14]))
  (cursor_blink_interval [12, 0] - [12, 32]
    interval: (number [12, 22] - [12, 24])
    ease: (ease [12, 25] - [12, 32]))
  (cursor_blink_interval [13, 0] - [13, 39]
    interval: (number [13, 22] - [13, 24])
    ease: (ease [13, 25] - [13, 39]
      (ease_step [13, 25] - [13, 39]
        count: (number [13, 31] - [13, 33])
        position: (ease_step_position [13, 35] - [13, 38]))))
  (mouse_shortcut [15, 0] - [15, 71]
    sequence: (mouse_sequence [15, 10] - [15, 14]
      (left [15, 10] - [15, 14]))
    event_type: (mouse_event [15, 15] - [15, 20])
    modes: (mouse_mode [15, 21] - [15, 30])
    action: (mouse_handle_click [15, 31] - [15, 71]
      actions: (handle_click_actions [15, 50] - [15, 71])))
  (mouse_shortcut [16, 0] - [16, 85]
    sequence: (mouse_sequence [16, 10] - [16, 20]
      (modified_key [16, 10] - [16, 20]
        (shift [16, 10] - [16, 15])
        (left [16, 16] - [16, 20])))
    event_type: (mouse_event [16, 21] - [16, 26])
    modes: (mouse_mode [16, 27] - [16, 44])
    action: (mouse_handle_click [16, 45] - [16, 85]
      actions: (handle_click_actions [16, 64] - [16, 85])))
  (mouse_shortcut [17, 0] - [17, 75]
    sequence: (mouse_sequence [17, 10] - [17, 25]
      (modified_key [17, 10] - [17, 25]
        (ctrl [17, 10] - [17, 14])
        (shift [17, 15] - [17, 20])
        (left [17, 21] - [17, 25])))
    event_type: (mouse_event [17, 26] - [17, 33])
    modes: (mouse_mode [17, 34] - [17, 51])
    action: (mouse_handle_click [17, 52] - [17, 75]
      actions: (handle_click_actions [17, 71] - [17, 75])))
  (mouse_shortcut [18, 0] - [18, 91]
    sequence: (mouse_sequence [18, 10] - [18, 29]
      (modified_key [18, 10] - [18, 29]
        (ctrl [18, 10] - [18, 14])
        (shift [18, 15] - [18, 20])
        (alt [18, 21] - [18, 24])
        (left [18, 25] - [18, 29])))
    event_type: (mouse_event [18, 30] - [18, 41])
    modes: (mouse_mode [18, 42] - [18, 59])
    action: (mouse_selection [18, 60] - [18, 91]
      selection: (mouse_selection_type [18, 76] - [18, 91])))
  (env [20, 0] - [20, 10]
    variable: (string [20, 4] - [20, 8])
    value: (string [20, 9] - [20, 10]))
  (env [21, 0] - [21, 26]
    variable: (string [21, 4] - [21, 8])
    value: (string [21, 9] - [21, 26])))
```

</details>

## Queries

Source of queries: "written from scratch"

<details>
<summary>Screenshots of code sample</summary>

<img width="1426" height="684" alt="tree-sitter-kitty" src="https://github.com/user-attachments/assets/db66283d-109a-43dc-81e3-481e5ee19485" />

</details>

<!--
CHECKLIST: _Before_ submitting, make sure

* `./scripts/install-parsers.lua <language>` works without warnings
* `./scripts/install-parsers.lua --generate <language>` works without warnings
* `make query` works without warning
* `make docs` is run
-->

clason commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,1001 @@
+; Command flags(`--flag`) should be highlighted *differently*.
+(kitten_arguments
+  (string) @variable.parameter
+  (#match? @variable.parameter "^-"))
```
`#lua-match?`

OXY2DEV replied:
I thought `#lua-match?` was a Neovim-specific thing. I was worried the parser would fail if it was loaded in a different editor.

I can change it if needed.

ribru17 replied:
Looks like this still needs to be changed

clason commented on runtime/queries/kitty/highlights.scm:
Honestly? It sounds like this might be a case where a simple syntax file is the much better option.

OXY2DEV replied:
> Sorry if this sounds dismissive of your efforts;

Nah, It's fine. It was mostly for fun. Who knows, maybe someone more knowledgeable than me would make a PR to fix it. Otherwise, I will just slowly do it myself(if I can).

ribru17 commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,1001 @@
+  path: (string) @string.special.path)
+
+(launch_hold_after_ssh
+  "--hold-after-ssh" @constant
```
Imo, most/all of these arguments should be aliased as `(arg)` or `(constant)` in the grammar, if possible; this would dramatically shrink query size. It's fine if the arguments *must* be this exact string to be considered an argument, but on the query side we don't really care about the literal text of them, just that they are arguments/constants, and should be highlighted as such.

ribru17 commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,1001 @@
+
+(launch_hold_after_ssh
+  "--hold-after-ssh" @constant
+  "="? @punctuation.delimiter
```
Looks like all `"="` are `@punctuation.delimiter`; it would be better just to have a top-level rule

```
"=" @punctuation.delimiter
```

This reduces query complexity a lot. (also just an fyi, it's usually better to break up one pattern with optional captures into just multiple patterns. this is usually more explicity/performant)

ribru17 commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,1001 @@
+(color_reset
+  "--reset" @constant
+  "="? @punctuation.delimiter
+  "no"? @boolean)
```
`"no"` could probably also just get its own top-level pattern

ribru17 commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,1001 @@
+  ] @variable.parameter)
+
+(mouse_mode
+  "," @punctuation.delimiter)
```
And here

ribru17 commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,1001 @@
+  name: (string) @string.special)
+
+; Comment ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+(comment) @comment
```
```suggestion
(comment) @comment @spell
```

ribru17 commented on runtime/queries/kitty/injections.scm:
"comment" injections as well, please :)

OXY2DEV commented:
@ribru17 

I have updated the queries.

I have,

- [x] Generalized a lot of the anonymous nodes into named nodes.
- [x] Moved `@punctuation` queries to top level9except ones that are node specific).
- [x] Replaced `"no"`s with boolean nodes.
- [x] Added injection for comments.
- [x] Added `@spell` to comments.

ribru17 commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,552 @@
+  value: (string) @string.special)
+
+;
+((titlebar_color) @constant
+  (#eq? @constant "system"))
+
+((titlebar_color) @constant
+  (#eq? @constant "background"))
```
there might be other opportunities for this refactor elsewhere... :thinking:

ribru17 commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,552 @@
+  characters: (string) @string.special)
+
+;
+(paste_action) @type
+
+;
+(pointer) @type
+
+;
+(layout) @type
```
nit: can also combine these into the other alternation

ribru17 commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,552 @@
+(cubic_bezier
+  "cubic-bezier" @function.call)
+
+(cubic_bezier
+  [
+    "("
+    ")"
+  ] @punctuation.bracket)
```
can also make these top level

ribru17 commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,552 @@
+(ease_step
+  "steps" @function.call)
+
+(ease_step
+  [
+    "("
+    ")"
+  ] @punctuation.bracket)
```
then we won't need this
```suggestion

```

ribru17 commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,552 @@
+((filter_sequence) @constant
+  (#eq? @constant "all"))
+
+(boolean_expression) @operator
```
If this looks like what I can see, that it is something like `"or"`, `"and"`, etc. then I think it should be `@keyword.operator`.

(nit: not a blocker, but I also think it would be clearer if it was named similar to that, since keywords/operators aren't expressions)

ribru17 commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,552 @@
+  (string) @string.regexp)
+
+(filter_element
+  "\"" @punctuation.delimiter)
```
typically the `"` should be a part of the `(string)` node, also captured as `@string`, and the inner content could then be e.g. `(string_content)`

ribru17 commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,552 @@
+  (string) @constant
+  (#any-of? @constant "auto" "monospace" "bold" "italic" "bold-italic") .)
+
+(font_property
+  name: (string) @variable.parameter
+  "=" @punctuation.delimiter)
```
```suggestion
(font_property
  name: (string) @variable.parameter)
```

ribru17 commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,552 @@
+  (string) @type)
+
+;
+(env
+  variable: (string) @variable.builtin
+  "=" @punctuation.delimiter)
```
```suggestion
(env
  variable: (string) @variable.builtin)
```

ribru17 reviewed (commented):
Thanks, colors looking good so far! Left more comments to help with refactoring, slightly more aggressive this time since the queries are a bit large still

OXY2DEV commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,552 @@
+  duration: (number) @number
+  suffix: (time_suffix) @type)
+
+;
```
The `duration` & `suffix` properties or the random `;`?

ribru17 replied:
Yeah the empty comments👍

ribru17 commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,467 @@
+
+; Color names & `none`.
+((color) @constant
+  (#match? @constant "^[^#]"))
```
Here as well

ribru17 commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,467 @@
+(launch_next_to
+  pattern: (string) @string.regexp)
+
+(launch_remote_password
+  [
+    "'"
+    "\""
+  ] @punctuation.delimiter)
+
+(launch_remote_password
+  password: (password) @string.special)
```
From the grammar's perspective, this should just be a `@string`, not a `@string.special`. It doesn't really have any semantic impact besides being a regular string, compared to e.g. if it denoted file metadata or interpreter version. I think it would be best to remove the quotation highlights and then just highlight the whole unit as `@string` for consistency. It would make sense, imo, to also capture the node as `(string)` (it already has the `password` field name anyway)

Also, I'm curious what you think: would it be feasible to rely on field names (or more on node names) for more generalized highlights? Many places in the grammar have highly specific node names which makes it difficult to write concise queries. Could we employ something similar to `(arg)` refactoring, but for general concepts like option names and builtin vars?

E.g. if we could do the following:

```
variable: (_) @variable.builtin
```

this would make the queries much shorter, helping a ton with maintainability, and they are more extensible as well. Once the grammar is updated to support a new kitty concept, the query automatically reflects highlights as long as it is given the `variable` field by the grammar. And so on for `type`, etc.

ribru17 commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,467 @@
+  path: (string) @string.special.path)
+
+(load_config_file
+  path: (string)? @string.special.path)
```
```suggestion
  path: (string) @string.special.path)
```

not necessary since this is the sole capture of the pattern. likewise with other patterns in the query

ribru17 commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,467 @@
+
+(scroll_to_prompt
+  prompt_number: (number) @number
+  lines_above: (number)? @operator)
```
Prefer breaking these two captures into two separate patterns for better performance (and then we won't need the `?`)

ribru17 reviewed (approved):
Thanks for the cleanup!

OXY2DEV commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,178 @@
+(flag) @constant
+
+((string) @constant
+  (#match? @constant "^-"))
```
Yeah, `tree-sitter highlight` doesn't like `#lua-match?` so I keep adding it back when testing and keep forgetting to revert it.

OXY2DEV commented on runtime/queries/kitty/highlights.scm:
```diff
@@ -0,0 +1,178 @@
+
+((color) @constant
+  (#lua-match? @constant "^[^#]"))
+
```
My bad. I forgot.
